### ISSUE TITLE
reducedim! and mapreducedim! aren't exported

### ISSUE THREAD

**Issue by diegozea** [JuliaLang/julia#19287]
`reducedim!` and `mapreducedim!` aren't exported. Is there a reason for that?
Best,

**Comment by nalimilan** [comment]
No longer relevant now that these have become methods of `reduce` and `mapreduce`.

**Comment by jekbradbury** [comment]
That’s true from the caller’s perspective, but in terms of an interface for array implementers I believe overloading `mapreduce` isn’t enough because some methods of functions like `sum` might still skip that and directly call `mapreducedim!`? What would be really nice is if an implementation could cover all required cases by overloading either of `mapreduce` or `mapreducedims!` (for instance, Tensorflow.jl would pick the former as it’s more functional and GPUArrays would pick the latter as it’s more imperative)

**Comment by nalimilan** [comment]
> I believe overloading mapreduce isn’t enough because some methods of functions like sum might still skip that and directly call mapreducedim!?

No, they don't (AFAICT, the code is hard to follow): they call `mapreduce`. Then you can override `mapreduce` at different points depending on you needs (notably whether your type is an `AbstractArray` or not).
